### PR TITLE
storage: partly fix doubled buckets

### LINKED ISSUE #414: Prevent doubled buckets caused by manual `bucket_send`
1. If manual vshard.storage.bucket_send() is used. Imagine this situation: storage S1 has bucket B. It is sent to S2, but then connection broke. The bucket is in the state S1 {B: sending-to-s2}, S2 {B: active}. Now if the user will do vshard.storage.bucket_send(S2 -> S3), then we will get this: S1 {B: sending-to-s2}, S2: {}, S3: {B: active}. Now when recovery fiber will wakeup on S1, it will see that B is sending-to-s2 but S2 doesn't have the bucket. Recovery will then assume that S2 already deleted B, and will recover it on S1. Now we have S1 {B: active} and S3 {B: active}. This is quite easy to achieve if one uses manual vshard.storage.bucket_send(). When automatic rebalancing is used, it shouldn't happen normally, because the rebalancer won't try to apply new routes while there are sending or receiving buckets anywhere.

2. The assumption in the last sentence in [1] might break probably due to master changes. Imagine that a replicaset had replicas R1 {= master} and R2. They lost connection between each other. R1 has a bucket which is sending/receiving but is already recovered to active state in another replicaset. Now R1 is demoted and R2 becomes master. It doesn't have this bucket in any state. And rebalancer will happily apply new routes. Then we might get the same situation as in [1]. Or something like that.

Maybe the fix would be as simple as to make vshard.storage.bucket_send() turn local bucket to SENT and only then the remote bucket to ACTIVE. Currently it is vice versa. Then [1] becomes not possible. If connection would be broken, then the local SENT bucket would be deleted eventually, and the remote RECEIVING would be turned to ACTIVE by recovery when the connection is back.

Proposed by @Gerold103 in https://github.com/tarantool/vshard/issues/412

### PR DESCRIPTION
**Pull request by Serpentian**
Currently buckets with 'active' status can appear on several shards in case of manual use of vshard.storage.bucket_send() in the following situation: storage S1 has bucket B, which is sent to S2, but then connection broke. The bucket is in the state S1 {B: sending-to-s2}, S2 {B: active}. Now if the user will do vshard.storage.bucket_send(S2 -> S3), then we will get this: S1 {B: sending-to-s2}, S2: {}, S3: {B: active}. Now when recovery fiber will wakeup on S1, it will see that B is sending-to-s2 but S2 doesn't have the bucket. Recovery will then assume that S2 already deleted B, and will recover it on S1. Now we have S1 {B: active} and S3 {B: active}, doubled buckets situation.

Let's fix that by turning local bucket to SENT and only after that the remote bucket to ACTIVE. It's safe to do, as local bucket will be garbade collected and the remote one will be turned to ACTIVE by recovery process.

Closes #414

NO_DOC=bugfix

### CODE REVIEW

File: test/rebalancer/receiving_bucket.test.lua [hunk 1]
```diff
@@ -43,6 +43,7 @@ box.space.test4:count()
 box.space.test5:count()
 
 vshard.storage.bucket_send(1, util.replicasets[2], {timeout = 10})
+wait_bucket_is_collected(1)
```
Comment (Serpentian) [inline test/rebalancer/receiving_bucket.test.lua:46]: This line is not related to the patch, it's just a possible flakiness (and it's reproduced). The bucket can be already garbage collected, let's fix it alongside.

**Comment by Serpentian** [comment]
The test `storage_1_1` fails on the master too, created ticked for that

File: test/instances/storage.lua [hunk 2]
```diff
@@ -120,6 +120,19 @@ local function bucket_gc_once()
     end)
```
Comment (olegrok) [inline test/instances/storage.lua:120]: typo in commit message: `garbade` -> `garbage`

**Comment by Serpentian** [comment]
I found a bunch of tests, which became flaky after that patch, I'll convert it to draft until they're fixed

File: test/instances/storage.lua (hunk 2 above)
Comment (Serpentian) [reply test/instances/storage.lua:120]: Fixed

File: test/luatest_helpers/vtest.lua [hunk 3]
```diff
@@ -71,8 +71,9 @@ end
 -- helps not to care.
 --
 local function error_is_timeout(err)
-    return err.type == 'ClientError' and err.code == box.error.TIMEOUT or
-           err.type == 'TimedOut'
+    return err.type == 'ClientError' and (err.code == box.error.TIMEOUT or
```
[the remark this answers is not in the document]
Comment (olegrok) [reply test/luatest_helpers/vtest.lua]: AFAIK this function was introduced due to changes in Tarantool and there are couple of modules that also contain such check. https://github.com/tarantool/cartridge/commit/73686c1cfd6a71c776ff2ebac125d295a9a3cde4#diff-aed3b7f006a0175c125574a5621fba5978870e2247cb6d719836f1386718062bR203-R208

So I assume that if this check is incorrect we could file issues to all repositories that contains such check.

[the remark this answers is not in the document]
Comment (Serpentian) [reply test/luatest_helpers/vtest.lua]: I changed this function, as in this particular case (due to inconsistency of timeout errors) returned value didn't have `box.error.TIMEOUT` as `err.code` in `ClientError`. The test didn't work on 2.7 and 2.8. AFAICS, there's no changes in these branches for almost a year. Moreover, I don't see any sense in changing a function in `lua-libs/util` as it works properly.

**Comment by Serpentian** [comment]
> Also please double-check yourself that my idea is feasible at all. Maybe I was wrong somewhere.

The test `storage_1_1_1` reproduces exactly what you said: without this patch we get doubled buckets.

File: vshard/storage/init.lua [hunk 4]
```diff
@@ -2113,6 +2114,17 @@ local function bucket_send_xc(bucket_id, destination, opts, exception_guard)
...
+    while M.errinj.ERRINJ_LAST_SEND_DELAY do
+        lfiber.sleep(0.01)
+    end
+    -- It is safe to turn bucket to SENT as it has already been fully
+    -- sent to the destination and can be garbade collected regardless of
+    -- receiving is_last message by the destination, as out there it will be
+    -- converted to ACTIVE status by recovery process. Moreover, it's preferable
+    -- to turn it into SENT at the source before turning to ACTIVE at the
+    -- destination, as it allows to avoid doubled buckets problem caused by
+    -- manual vshard.storage.bucket_send() call (gh-414).
```
[the remark this answers is not in the document]
Comment (Serpentian) [reply vshard/storage/init.lua]: Done

File: test/storage-luatest/storage_1_1_1_test.lua [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+-- doubled buckets (gh-414).
+--
+test_group.test_manual_bucket_send_doubled_buckets = function(g)
+    vtest.cluster_exec_each_master(g, function()
+        _G.bucket_recovery_pause()
+    end)
+
+    local uuid_2 = g.replica_2_a:exec(function()
+        ivshard.storage.internal.errinj.ERRINJ_LONG_RECEIVE = true
+        return box.info.cluster.uuid
```
Comment (Gerold103) [inline test/storage-luatest/storage_1_1_1_test.lua:70]: This won't work in 3.0 - you need to rebase on latest vshard master and use `util.replicaset_uuid()`.

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the fixes! Only nits are left.

File: test/instances/storage.lua [hunk 6]
```diff
@@ -120,6 +120,19 @@ local function bucket_gc_once()
...
 end
 
+local function bucket_recovery_wait()
+    local status_index = box.space._bucket.index.status
+    t.helpers.retrying({timeout = wait_timeout}, function()
+        vshard.storage.garbage_collector_wakeup()
```
[the remark this answers is not in the document]
Comment (Serpentian) [reply test/instances/storage.lua:126]: I should. Fixed

File: test/rebalancer/bucket_ref.result [hunk 7]
```diff
@@ -124,23 +124,29 @@ vshard.storage.buckets_info(1)
...
+f2 = fiber.create(function()
+    vshard.storage.bucket_send(1, util.replicasets[2])
+end)
+while true do
+    local i = vshard.storage.buckets_info(1)[1]
+    if i.status == vshard.consts.BUCKET.SENDING then
+        break
+    end
+    fiber.sleep(0.01)
+end;
```
[the remark this answers is not in the document]
Comment (Serpentian) [reply test/rebalancer/bucket_ref.result]: Fixed

**Review by Gerold103 — APPROVED** [review]
Thanks for all the work!
